docs/architecture/how/meta-tx.md

```diff
@@ -0,0 +1,77 @@
+# Meta Transactions
+
+[NEP-366](https://github.com/near/NEPs/pull/366) introduced the concept of meta
+transactions to Near Protocol. The idea is to construct and sign a transaction
+off-chain and allow someone else to release it on the chain. This someone else
+is called the relayer, who will then pay for the gas and tokens costs of the
+transaction. This ultimately allows executing transactions on NEAR without
+owning gas or tokens.
+
+## Overview
+
+![Flow chart of meta
+transactions](https://raw.githubusercontent.com/near/NEPs/003e589e6aba24fc70dd91c9cf7ef0007ca50735/neps/assets/nep-0366/NEP-DelegateAction.png)
+_Credits for the diagram go to the NEP authors Alexander Fadeev and Egor
+Uleyskiy._
+
+
+TODO: Describe the flow chart
+
+## Gas costs for meta transactions
+
+Meta transactions challenge the traditional ways of charging gas for actions. To
+see why, let's first list the normal flow of gas, outside of meta transactions.
+
+1. Gas is purchased by the transaction signer when it is converted to a receipt.
+   This happens as part of `verify_and_charge_transaction` which gets called in
+   `process_transaction`.
+2. For all actions listed inside the transaction, the `SEND` cost is burned
+   immediately. Depending on the condition `sender == receiver`, one of two
+   possible `SEND` costs is chosen. The `EXEC` cost is not burned, yet. But it
+   is implicitly part of the transaction cost. The third and last part of the
+   transaction cost is the gas attached to function calls. The attached gas is
+   also called prepaid gas. (Not to be confused with `total_prepaid_exec_fees`
```

Comment:
Hm, yes, you can think of them as these 3 logical fields in this context. But you don't find them stored exactly like that. And throughout the code, there are more fields and their names are somewhat inconsistent...

If you look at `ActionResult`, the actual gas fields stored are:
```rust
pub gas_burnt: Gas,
pub gas_burnt_for_function_call: Gas,
pub gas_used: Gas,
```
Here the `gas_used` is = `burned_gas + total_prepaid_exec_fees + prepaid_gas(outgoing)`.

The values for `total_prepaid_exec_fees` and `prepaid_gas` are never stored explicitly on the receipt level. It's always implicitly defined by the list of actions.

`prepaid_gas` is only relevant for function calls and 0 everywhere else. Adding up the prepaid gas inn all function calls of the receipt will give you the total `prepaid_gas`.

Likewise, the `total_prepaid_exec_fees` is the sum of all execution fees for all actions. We don't have to store it explicitly, because we can recompute it if we know the list of actions, which is why I wrote it is "implicitly" part of the total cost.

`gas_burnt_for_function_call` is not relevant for this discussion, it is only tracked separately for the 30% smart contract reward that only applies to the fraction of the cost that was burned for the actual function call.

In `ExecutionOutcome` you only have one gas field:
```rust
pub gas_burnt: Gas,
```
Everything else is implicitly defined by looking at the outgoing receipts.